### PR TITLE
Editor: Add WordPress.com integrations to the block editor

### LINKED ISSUE #32044: Gutenframe: Align the iframe bridge communication code for both Jetpack and Simple sites.
This involves moving the active WP.com file to JP and sync it with Fusion, so that the code already correcting issues for WP.com can also be applied to Jetpack sites (media modal, revisions, back button, etc.).

See: p7jreA-2gu-p2.
Background: p3fqKv-6Hh-p2.

### PR DESCRIPTION
**Pull request by mmtr**
Fixes https://github.com/Automattic/wp-calypso/issues/32044.
Background: p3fqKv-6Hh-p2.

### Changes proposed in this Pull Request:

Load the scripts of the [WordPress.com block editor integrations](https://github.com/Automattic/wp-calypso/tree/master/apps/wpcom-block-editor) that enhances the block editor with the features available in WordPress.com sites

**Calypso Iframed block editor:**
- URL is updated with the draft ID when the post is saved.
- Trashing a post redirects to the list of trashed posts in Calypso displaying a success notification.
- Revisions are managed with the Calypso revisions modal.
- Reblogging a post from the Calypso reader creates a post in Calypso with the except title and content.
- Media items are managed with the Calypso media modal.
- Posts are previewed with the Calypso preview modal.
- Back button on the header ("View posts") redirects to the list of posts in Calypso.
- Media items inserted in classic blocks are managed with the Calypso media modal.

**WP Admin block editor** (also included in the iframed block editor):
- Support for the underline and justified text formats.

#### Testing instructions:
* Start Calypso locally using the `update/me-sites-version` branch (https://github.com/Automattic/wp-calypso/pull/32527).
* Go to `/block-editor/post/:yourJetpackDomain`.
* Verify all the integrations listed above work as expected.
* Load `:yourJetpackDomain/wp-admin/post-new.php`.
* Check the underline and justified text formats are available on the paragraph blocks. 

#### Proposed changelog entry for your changes:
I guess it's not needed?

### CODE REVIEW

File: modules/wordpress-com-block-editor-iframe.php
Comment (jeherve) [inline modules/wordpress-com-block-editor-iframe.php]: You'll probably want to add the textdomain here and elsewhere.

File: modules/wordpress-com-block-editor-iframe.php
Comment (jeherve) [inline modules/wordpress-com-block-editor-iframe.php]: Maybe we could only enqueue it on the pages where we need it? I think that would be nice given the number of dependencies we are calling here.

**Review by jeherve — CHANGES_REQUESTED** [review]
Quick first pass.

File: modules/wordpress-com-block-editor-iframe.php
Comment (obenland) [reply modules/wordpress-com-block-editor-iframe.php]: @mmtr To save you some leg work, `admin_enqueue_scripts` passes `$hook_suffix` that we can test for `post.php` and `post-new.php`. I believe those are the two pages for editing an existing post and creating a new one.

Alternatively we could also use `enqueue_block_editor_assets`, if that is not too late in the execution path.

Comment (kwight) [reply modules/wordpress-com-block-editor-iframe.php]: I'd agree with @jeherve . I'm not sure `get_current_screen()` is reliable yet at this point, but I think a `$pagenow` check for either `post.php` or `post-new.php` would work anyways.

File: modules/wordpress-com-block-editor-iframe.php
Comment (kwight) [inline modules/wordpress-com-block-editor-iframe.php]: The callback is misspelled here, causing `call_user_func_array()` log warnings and the plugin not to be loaded:

```suggestion
	add_filter( 'mce_external_plugins', 'jetpack_add_wpcom_block_editor_tinymce_plugins' );
```

File: modules/wordpress-com-block-editor-iframe.php
Comment (mmtr) [reply modules/wordpress-com-block-editor-iframe.php]: Using `enqueue_block_editor_assets` did the trick. Thanks @obenland for the suggestion!

**Comment by jeherve** [comment]
Looks like this needs a rebase now.

File: modules/wordpress-com-block-editor-iframe.php
Comment (dmsnell) [inline modules/wordpress-com-block-editor-iframe.php]: we should create an issue or some notice to come back and change this at some point. in my past I've regretted forgetting to remove date-based cache-busters.

File: modules/wordpress-com-block-editor-iframe.php
Comment (dmsnell) [inline modules/wordpress-com-block-editor-iframe.php]: this might hurt our ability to find the script in the future. we might consider creating a bit of duplication in order to preserve searchability

```php
$use_debug = defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG;
$path = $use_debug
	? '//widgets.wp.com/wpcom-block-editor/common.min.js'
	:  '//widgets.wp.com/wpcom-block-editor/common.js';
```

File: modules/wordpress-com-block-editor-iframe.php
Comment (dmsnell) [inline modules/wordpress-com-block-editor-iframe.php]: sam comment as above concerning search ability of the script path

File: modules/wordpress-com-block-editor-iframe.php
Comment (dmsnell) [inline modules/wordpress-com-block-editor-iframe.php]: _ibid_ searchability

File: modules/wordpress-com-block-editor-iframe.php
Comment (mmtr) [reply modules/wordpress-com-block-editor-iframe.php]: What problems did you have with versions based on dates? The reasoning behind this approach is to make sure that Jetpack users are getting the latest version of the scripts without requiring them to update the plugin (which also removes us the necessity to maintain this code every time we update the scripts).

Do you have in mind any other mechanism not involving date-based cache-busters that also maintains the dynamic versioning flexibility?

Comment (dmsnell) [reply modules/wordpress-com-block-editor-iframe.php]: the main consequences were that once the code reached a stable point people were needlessly re-downloading the script.

in this case a week isn't so bad but I just wanted to somehow mark this for a follow-up in something like six months to see if it's still needed

Comment (mmtr) [reply modules/wordpress-com-block-editor-iframe.php]: Thanks for the context. It makes sense :) I created a [temporary card](https://github.com/orgs/Automattic/projects/88#card-20578307) in the House Tarly board which I'll convert to an issue once this PR is merged.

**Comment by obenland** [comment]
@jeherve Could you give this another look when you get a chance?

**Comment by kwight** [comment]
Note that this should be rebased again and approved after #11354 is merged (it was based on #11354 for easier testing).

File: modules/wpcom-block-editor/class-jetpack-wpcom-block-editor.php
Comment (jeherve) [inline modules/wpcom-block-editor/class-jetpack-wpcom-block-editor.php]: Should this be enqueued when Calypsoify is not active?

**Review by jeherve — COMMENTED** [review]
I can't test in `wpcalypso` because of mixed content errors, but I could test locally.

Everything seemed to work, but I could see quite a few JavaScript errors in the Calypso editor, and some in the wp-admin editor as well:

image

Those may not be related to this PR though.

File: modules/wpcom-block-editor/class-jetpack-wpcom-block-editor.php
Comment (mmtr) [reply modules/wpcom-block-editor/class-jetpack-wpcom-block-editor.php]: At least is what we do in WordPress.com but I'm not sure we have to do it here. The `wpcom-block-editor` scripts provide integrations for two environments: Calypso iframed block editor and WP Admin block editor ([#](https://github.com/Automattic/wp-calypso/tree/master/apps/wpcom-block-editor#environments)).

The `wpcom-block-editor/calypso-*.js` scripts enhances the Calypso iframed block editor replacing core tools like media modal or revisions which Calypso tools, which is why we only load when Calypsoify is active (my last commit changes this check in favor of the new `is_iframed_block_editor` method to make clear is not Calypsoify but the iframed block editor).

The `wpcom-block-editor/common.js` scripts enhances both editor, which is why we load them always. These enhancements add some missing features in Core like justify or underline text formats, or the ability to switch to the classic editor from the block editor. I think it makes sense to have these features on a Jetpack site connected to WordPress.com, so users will have the same experience in both envs.

**Comment by mmtr** [comment]
> Everything seemed to work, but I could see quite a few JavaScript errors in the Calypso editor, and some in the wp-admin editor as well:
> 
> image
> 
> Those may not be related to this PR though.

Thanks for testing! I can see the same errors while testing the WP Admin block editor on Jetpack master. The Calypso JS errors are also present on my local environment while testing other WordPress.com sites. So definitely not related to this PR

**Review by jeherve — APPROVED** [review]
Looking good now, although I think this could benefit from a rebase, some of those commits should not be here.

File: modules/wordpress-com-block-editor-iframe.php
Comment (noahtallen) [reply modules/wordpress-com-block-editor-iframe.php]: > we should create an issue or some notice to come back and change this at some point. in my past I've regretted forgetting to remove date-based cache-busters.

@dmsnell you were totally right....... 

This week a version of the script was deployed which contained a bug. It was only noticed after the cache rolled over to the next day. So we reverted the faulty code and deployed it. But then people were confused why it wasn't fixed, because it took another 24 hours for the fix to be available to all users because of the cache.

So yeah, we definitely need to change this to something which invalidates automatically when the file is updated. 🤔

**Pull request merged**
